kaldb/src/main/java/com/slack/kaldb/preprocessor/PreprocessorService.java

```diff
@@ -268,40 +262,29 @@ protected static List<Integer> getActivePartitionList(DatasetMetadata datasetMet
    * valid dataset metadata are provided throws an exception.
    */
   protected static StreamPartitioner<String, Trace.Span> streamPartitioner(
-      Map<String, List<Integer>> datasetToPartitionList) {
-    checkArgument(
-        datasetToPartitionList.entrySet().size() > 0, "datasetToPartitionList cannot be empty");
-    checkArgument(
-        datasetToPartitionList.keySet().stream().noneMatch(String::isEmpty),
-        "datasetToPartitionList cannot have any empty keys");
+      DatasetMetadata datasetMetadata) {
+    //    checkNotNull(datasetMetadata, "dataset cannot be null");
+    checkArgument(datasetMetadata != null, "dataset cannot be null");
     checkArgument(
-        datasetToPartitionList.values().stream().noneMatch(List::isEmpty),
+        getActivePartitionList(datasetMetadata).size() > 0,
         "datasetToPartitionList cannot have any empty partition lists");
 
-    return (topic, key, value, partitionCount) -> {
-      String datasetName = PreprocessorValueMapper.getDatasetName(value);
-      if (!datasetToPartitionList.containsKey(datasetName)) {
-        // this shouldn't happen, as we should have filtered all the missing datasets in the value
-        // mapper stage
-        throw new IllegalStateException(
-            String.format("Dataset '%s' was not found in dataset metadata", datasetName));
-      }
-
-      List<Integer> partitions = datasetToPartitionList.getOrDefault(datasetName, List.of());
+    return (topic, key, value, numPartitions) -> {
+      List<Integer> partitions = getActivePartitionList(datasetMetadata);
```

Comment:
this also has the advantage that we aren't deserializing the record for the 2nd time - perhaps improving efficiency